fix sort algorithm: singleton vs sorted buckets

bruno-at-orange opened the pull request:
The bug was introduced in version 10.3.1. This occurs when there are singleton buckets (also known as 'elephant keys') and the same separator is used in the input and output files.

bruno-at-orange committed:
fix sort algorithm: singleton vs sorted buckets

bucket->GetSorted() return false if bucket is a singleton (new behavior in 10.3.1)
Buckets must be concatenated if they are sorted OR singletons.

marcboulle reviewed (commented):
En relisant la méthode KWChunkSorterTask::MasterFinalize, j'ai quelques questions sur la clarification du lien entre GetOutputFileName et GetChunkFileNames par bucket.

bruno-at-orange committed:
Remove files in KWChunkSorterTask in case of errors (previously done outside the task)

bruno-at-orange committed:
Fix GetSortedLineNumber() in KWChunkSorterTask

Line number is computed during the sort in the SlaveProcess. Singleton chunks, are not processed by SlaveProcess unless the separators are different.
Therefore, lines are not counted. We now increment the lines of the singleton in the MasterPrepareInput.

Lines number is now aggregated in KWSortedChunkBuilderTask the built buckets contains the size of the bucket and the line number
Minor refactoring in KWSortedChunkBuilderTask, we no longer use an ObjectDictionary to store and access buckets,
instead we access the buckets directly from the shared object array.

marcboulle reviewed (changes requested):
Améliorer le commentaire de KWSortBucket::AddLine
Ajouter une méthode de type GetTotalLineNumber()?

Peut-on compter les ligne des buckets singletons dans les esclaves plutôt que dans le master?
Sinon, et plutôt, est-il possible de compter les lignes des chunks au fur et à mesure de leur alimentation (via le AddLine?)

marcboulle reviewed (approved):
Ce me parait bien dans l'ensemble, mais je n'ai pas pu relire dans les détail, compte tenu du nombre de changements.
Je met en approuve pour ne pas bloquer le merge, mais j'ai quand même quelques question sur le nettoyage des fichiers en fin de tâche, à mieux commenter, et à clarifier en cas d'interruption utilisateur.
